x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/Decay.java

```diff
@@ -0,0 +1,602 @@
+
+        EvalOperator.ExpressionEvaluator.Factory decayFactory = decay != null
+            ? toEvaluator.apply(decay)
+            : EvalOperator.DoubleFactory(DEFAULT_DECAY);
+
+        EvalOperator.ExpressionEvaluator.Factory typeFactory = type != null
+            ? toEvaluator.apply(type)
+            : EvalOperator.BytesRefFactory(new BytesRef(DEFAULT_FUNCTION));
+
+        return switch (valueDataType) {
+            case INTEGER -> new DecayIntEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case DOUBLE -> new DecayDoubleEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case LONG -> new DecayLongEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case GEO_POINT -> new DecayGeoPointEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case CARTESIAN_POINT -> new DecayCartesianPointEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case DATETIME -> new DecayDatetimeEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            case DATE_NANOS -> new DecayDateNanosEvaluator.Factory(
+                source(),
+                valueFactory,
+                originFactory,
+                scaleFactory,
+                offsetFactory,
+                decayFactory,
+                typeFactory
+            );
+            default -> throw new UnsupportedOperationException("Unsupported data type: " + valueDataType);
+        };
+    }
+
+    @Evaluator(extraName = "Int")
+    static double process(int value, int origin, int scale, int offset, double decay, BytesRef functionType) {
+        return switch (functionType.utf8ToString()) {
+            case "exp" -> new ScoreScriptUtils.DecayNumericExp(origin, scale, offset, decay).decayNumericExp(value);
```

Comment:
For numeric Decay I simply reused `ScoreScriptUtils`, which are also used by painless